### ISSUE TITLE
clearScroll causes 403 when using elasticsearch 12.1.1 and higher

### ISSUE THREAD

**Issue by ajkerr** [TheDeveloper/http-aws-es#19]
Something changed in the elasticsearch module in version 12.1.1 and up, which is causing request signing errors for the clearScroll() method.

"The request signature we calculated does not match the signature you provided."

Reverting to 12.1.0 fixes the issue.

I think that this commit is the culprit: https://github.com/elastic/elasticsearch-js/commit/97706ce5fdd4fbdf4f00a9d5e6da0b3090764d33

**Comment by jonberke** [comment]
We are experiencing the problem but can't downgrade to 12.1.0 of the ES library because it doesn't support the ES version we are running (5.3).

We've contacted AWS support about this but they seem to blame this module for not generating the signature correctly.  I can't see how this module can incorrectly generate the signature for the clearScroll request but generate it correctly for every other request.

Has anyone found a workaround for this problem that works with ES 5.3 or do you have any information I can pass to AWS to get them to look into this issue more deeply?

**Comment by ajkerr** [comment]
@jonberke We still have the problem, it it's becoming more serious because it prevents us from moving to a newer version of Elasticsearch hosted in AWS. We are currently stuck on 2.x.

I've written a couple of sample programs that show that illustrate the problem:

This first program is written for http-aws-es 1.1.3 and elasticsearch 12.1.0. I was unable to use newer versions of http-aws-es here because of a peerDependency issue.

```javascript
const elasticsearch = require("elasticsearch");
const AWS = require("aws-sdk");

AWS.config.getCredentials((err) => {
  if (err) {
    console.log(`[error] credentials are not available: ${JSON.stringify(err)}`);
  }
  const amazonESProperties = {
    region: "us-west-2",
    credentials: AWS.config.credentials,
  };
  const client = new elasticsearch.Client({
    host: "https://your-es-endpoint.amazonaws.com",
    connectionClass: require("http-aws-es"),
    amazonES: amazonESProperties,
    apiVersion: "2.4",
    log: "debug"
  });

  client.clearScroll({scrollId: "12345"}, (err, res, status) => {
    if (err || status !== 200) {
      console.log(`${status} status. Could not clear scroll: ${JSON.stringify(err)}`);
    }
    console.log("res: %j", res);
  });
});
```

This program will fail with `400 status. Could not clear scroll`, which is correct, because the scroll id is made up.

The second program is written for http-aws-es 3.1.0 and elasticsearch 13.3.1 (the latest at the time of writing):

```javascript
const elasticsearch = require("elasticsearch");
const AWS = require("aws-sdk");

AWS.config.update({region: "us-west-2"});

const client = new elasticsearch.Client({
  host: "https://your-es-endpoint-es.amazonaws.com",
  connectionClass: require("http-aws-es"),
  apiVersion: "2.4",
  log: "debug"
});

client.clearScroll({scrollId: "12345"}, (err, res, status) => {
  if (err || status !== 200) {
    console.log(`${status} status. Could not clear scroll: ${JSON.stringify(err)}`);
  }
  console.log("res: %j", res);
});
```

In this case, it fails with a 403 status:

```
403 status. Could not clear scroll: {"msg":"Authorization Exception","path":"/_search/scroll","query":{},"body":"{\"scroll_id\":\"12345\"}","statusCode":403,"response":"{\"message\":\"The request signature we calculated does not match the signature you provided. Check your AWS Secret Access Key and signing method. Consult the service documentation for details.
```

As I mentioned when I first opened the bug, this started happening with version 12.1.1 of the elasticsearch module.

**Comment by ajkerr** [comment]
Note that calling scroll() instead of clearScroll() works fine. I can't see must difference between the generated requests other than the HTTP method. Perhaps it is a bug in the AWS signing code?

**Comment by ajkerr** [comment]
@jonberke I'm leaning towards the latter. I found this old forum thread: https://forums.aws.amazon.com/thread.jspa?threadID=227353

**Comment by ajkerr** [comment]
I've opened https://github.com/aws/aws-sdk-js/issues/1733 to track this issue as well. I created a sample program that shows the issue only using the aws-sdk.

**Comment by jonberke** [comment]
Awsome!  I'll include a link to that issue in my case with AWS support.

**Comment by jonberke** [comment]
Good news.  Got this from AWS support:

> Elasticsearch Service Team identified the root cause of the problem. The problem is with the aws-sdk-js library only.
> 
> When we send the DELETE with body, body is not sent in the request, causing the signature mismatch.
> 
> FYI - Elasticsearch Service team is doing follow-up with the sdk team regarding this.

@ajkerr I think your script with only the aws sdk did the trick.  Nicely done.

**Comment by ajkerr** [comment]
@jonberke That's good to hear. Assuming that they fix the JS SDK in a timely manner, the only fix for this repo would be to update the peerDependencies in package.json to the version of the SDK with the fix.

**Comment by TheDeveloper** [comment]
Thanks for digging into this @ajkerr & @jonberke!

Happy to bump the peerDeps when fix is live. Assuming they push it as patch it should match the semver anyway.

**Comment by ajkerr** [comment]
@jonberke @TheDeveloper I've added some more info to https://github.com/aws/aws-sdk-js/issues/1733 that should help them fix the problem. Adding a proper Content-Length header to the request seems to fix the issue.

**Comment by subodhkhanduri1** [comment]
The request body is not sent because of the missing Content-Length header.

I've opened https://github.com/TheDeveloper/http-aws-es/pull/41 to fix this. Please review.

**Comment by jonberke** [comment]
@TheDeveloper I was just wondering if you had an ETA for when you expect to merge the pull request that fixes this issue? Thanks.

**Comment by ajkerr** [comment]
@TheDeveloper I tend to agree that the submitted PR is the best best for fixing this, based on comments in https://github.com/aws/aws-sdk-js/issues/1733#issuecomment-336262403

**Comment by jonberke** [comment]
@TheDeveloper Any chance we can get this PR integrated soon?

**Comment by TheDeveloper** [comment]
Great work. Sorry for the delay. Merged in v3.1.1
